client/CServerHandler.cpp

```diff
@@ -245,88 +256,97 @@ void CServerHandler::startLocalServerAndConnect()
-	while(!c && state != EClientState::CONNECTION_CANCELLED)
-	{
-		try
-		{
-			logNetwork->info("Establishing connection...");
-			c = std::make_shared<CConnection>(
-					addr.size() ? addr : getHostAddress(),
-					port ? port : getHostPort(),
-					NAME, uuid);
-
-			nextClient = std::make_unique<CClient>();
-			c->iser.cb = nextClient.get();
-		}
-		catch(std::runtime_error & error)
-		{
-			logNetwork->warn("\nCannot establish connection. %s Retrying in 1 second", error.what());
-			boost::this_thread::sleep_for(boost::chrono::milliseconds(1000));
-		}
-	}
+	logNetwork->info("Establishing connection to %s:%d...", addr, port);
+	setState(EClientState::CONNECTING);
+	serverHostname = addr;
+	serverPort = port;
 
-	if(state == EClientState::CONNECTION_CANCELLED)
+	if (!isServerLocal())
 	{
-		logNetwork->info("Connection aborted by player!");
-		return;
+		Settings remoteAddress = settings.write["server"]["remoteHostname"];
+		remoteAddress->String() = addr;
+
+		Settings remotePort = settings.write["server"]["remotePort"];
+		remotePort->Integer() = port;
 	}
 
-	c->handler = std::make_shared<boost::thread>(&CServerHandler::threadHandleConnection, this);
+	networkHandler->connectToRemote(*this, addr, port);
+}
+
+void CServerHandler::onConnectionFailed(const std::string & errorMessage)
+{
+	assert(getState() == EClientState::CONNECTING);
+	boost::mutex::scoped_lock interfaceLock(GH.interfaceMutex);
 
-	if(!addr.empty() && addr != getHostAddress())
+	if (isServerLocal())
 	{
-		Settings serverAddress = settings.write["server"]["server"];
-		serverAddress->String() = addr;
+		// retry - local server might be still starting up
+		logNetwork->debug("\nCannot establish connection. %s. Retrying...", errorMessage);
+		networkHandler->createTimer(*this, std::chrono::milliseconds(100));
 	}
-	if(port && port != getHostPort())
+	else
 	{
-		Settings serverPort = settings.write["server"]["port"];
-		serverPort->Integer() = port;
+		// remote server refused connection - show error message
+		setState(EClientState::NONE);
+		CInfoWindow::showInfoDialog(CGI->generaltexth->translate("vcmi.mainMenu.serverConnectionFailed"), {});
 	}
 }
 
-void CServerHandler::applyPacksOnLobbyScreen()
+void CServerHandler::onTimer()
 {
-	if(!c || !c->handler)
-		return;
+	boost::mutex::scoped_lock interfaceLock(GH.interfaceMutex);
 
-	boost::unique_lock<boost::recursive_mutex> lock(*mx);
-	while(!packsForLobbyScreen.empty())
+	if(getState() == EClientState::CONNECTION_CANCELLED)
 	{
-		boost::mutex::scoped_lock interfaceLock(GH.interfaceMutex);
-		CPackForLobby * pack = packsForLobbyScreen.front();
-		packsForLobbyScreen.pop_front();
-		CBaseForLobbyApply * apply = applier->getApplier(CTypeList::getInstance().getTypeID(pack)); //find the applier
-		apply->applyOnLobbyScreen(dynamic_cast<CLobbyScreen *>(SEL), this, pack);
-		GH.windows().totalRedraw();
-		delete pack;
+		logNetwork->info("Connection aborted by player!");
+		return;
 	}
+
+	assert(isServerLocal());
```

Comment:
This code should only be reachable in case of local server (see method above - timer is only created for local server)
Also next line makes unconditional connection attempt to *local* server, so if we somehow end up here with remote server vcmi will actually try to connect to local server anyway.